src/lib/ingest.js

```diff
@@ -1,53 +1,264 @@
+      try {
+        await this.config.successHandler(item)
+      } catch (err) {
+        logger.error(`Ingest success hander failed: ${err}`)
+      }
+    }
+  }
+
+  async errorHandler(item, error) {
+    if (error instanceof Error) {
+      error = error.toString()
+    } else if (error instanceof String) {
+      // pass
+    } else {
+      error = JSON.stringify(error)
+    }
+    logger.error(`Error ingesting ${item.id}: ${error}`)
+    if (this.config.errorHandler) {
+      try {
+        await this.config.errorHandler(item, error)
+      } catch (err) {
+        logger.error(`Ingest error hander failed: ${err}`)
+      }
+    }
+  }
+
+  async ingestCollection(collection) {
+    try {
+      const index = COLLECTIONS_INDEX
+      const id = collection.id
+      const body = this.collectionToEs(collection)
+      await this.client.update({
+        index,
+        type: '_doc',
+        id,
+        body
+      })
+
+      logger.info(`Ingested collection ${id}; attempting to create index`)
+      await esClient.createIndex(id)
+      logger.info(`Index created for collection ${id}`)
+      this.collections.add(id)
+      await this.successHandler(this.addLinks(collection, body.doc))
+    } catch (error) {
+      await this.errorHandler(collection, error)
+    }
+  }
+
+  async ingestItems(items) {
+    const promises = []
+
+    // prepare items for ES query
+    const itemsById = {}
+    const operations = []
+    items.forEach((item) => {
+      try {
+        const { operation, body } = this.itemToEs(item)
+        itemsById[item.id] = {
+          original: item,
+          converted: body.doc
+        }
+        operations.push(operation)
+        operations.push(body)
+      } catch (err) {
+        promises.push(this.errorHandler(item, err))
+      }
+    })
+
+    if (operations.length === 0) {
+      logger.warn('No items remaining to be ingested')
+      return
+    }
+
+    logger.debug(`Writing batch of documents size ${operations.length / 2}`)
+    // run bulk ES query with items
+    await this.client.bulk({ body: operations }).then(async (response) => {
+      const esRespItems = response.body.items
+      logger.debug(`Wrote batch of documents size ${operations.length / 2}`)
+
+      // process ES response, handling successes and errors
+      let success = 0
+      esRespItems.forEach((esRespItem) => {
+        const esItem = esRespItem[(Object.keys(esRespItem)[0])]
+        const { original, converted } = itemsById[esItem._id]
+        if (esItem.error) {
+          promises.push(this.errorHandler(original, esItem.error))
         } else {
-          logger.debug('Ingested item')
-          resolve(true)
+          success += 1
+          promises.push(this.successHandler(this.addLinks(original, converted)))
         }
+      })
+      await Promise.allSettled(promises).then(() => {
+        logger.info(`${success} of ${items.length} items successfully ingested`)
+      })
+    })
+  }
+
+  async ingest(records) {
```

Comment:
This implementation could make it easier to understand what's happening here, and in what order:

```js
async ingest(records) {
  const [collections, otherItems] = partition(records, stacUtils.isCollection)

  // Ingest collections first
  await Promise.allSettled(collections.map((c) => this.ingestCollection(c)))

  const [items, invalidItems] = partition(otherItems, stacUtils.isItem)

  // Execute error handlers for invalid items
  const invalidItemHandlers = invalidItems.map(
    (i) => this.errorHandler(i, 'Unable to determine record type')
  )

  // Wait for items to be ingested and error handlers to run
  await Promise.allSettled([
    this.ingestItems(items),
    ...invalidItemHandlers
  ])
}
```